docs/madr/decisions/020-kds-redesign.md

```diff
@@ -0,0 +1,179 @@
+
+Tried:
+- upgrade path
+- rollback
+
+It looks that state returns to the previous one and nothing breaks.
+
+### Handshake
+
+Is it possible for the zone control-plane to be newer than the global control-plane?
+
+If the answer is no, then a handshake is not necessary as the client will use the latest available version.
+
+If the answer is yes, then a handshake API may need to be implemented. The GRPC Reflection API allows services to be discovered on the server side, but not on the client side.
+
+The output of the reflection functionality is:
+
+```
+["grpc.reflection.v1alpha.ServerReflection", "kuma.mesh.v1alpha1.GlobalKDSService", "kuma.mesh.v1alpha1.MultiplexService"]```
+```
+
+How would a handshake work?
+
+1. The server exposes the RPC `handshake`.
+2. The client initiates a connection to the server.
+3. The client sends a request that includes the Kuma version and the supported KDS versions:
+```
+ kumaVersion: "2.1.0",
+ supportedKdsVersions: ["0.1.0", "0.2.0"],
+```
+4. Server checks the request and responds with the newest version working for both
+```
+ kdsVersion: "0.2.0"
+```
+5. The client needs to create RPC clients for the specific version.
+
+### Multitenant
+
+Global control-plane should support multitenant, that means we want to achieve that one global control-plane can handle many independent zone control-planes. One approach to accomplish this is by extracting tenant specifics from an authentication token. These details can then be transmitted along with a request to indicate the specific tenant and retrieve solely their deployment information. Another option is to supply this information via a header. It appears that accommodating multitenancy doesn't require additional work on the KDS side.
+
+### Incremental xDS
+
+Incremental xDS in the `DeltaDiscoveryRequest` has a field `resource_names_subscribe` which requires information about resources that we want to subscribe to e.g. policy with name `timeout-1`. If you send `resource_names_subscribe:  "*"` in a `DeltaDiscoveryRequest`, the control plane should respond with all resources of the specified type, not just the ones that have changed since the last update. This is because the `resource_names_subscribe` field is used to filter which resources the control plane should send, and if it is set to `*`, it effectively disables filtering and requests all resources of the specified type. Looks like go-control-plane has no implementation of the client, so it might requires us to implement one.
+
+xDS Specification
+* https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#how-the-client-specifies-what-resources-to-return
+* https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#incremental-xds
+* https://github.com/envoyproxy/go-control-plane/blob/main/pkg/cache/v3/delta.go#L31
+
+### Measure performance of current KDS solution
+
+As a part of a task we could consider mesuring performance of the KDS to compare it with the new solution.
```

Comment:
Will we do this _before_ deciding to redesign KDS or will we do it after we've implemented a redesign?

Reply:
This wasn't the motivation for the redesign. The reason is to simplify and separate bi-directional streaming and make it more readable. Also, changing the way how we synchronize resources to don't send all the state.

Reply:
Should we just leave it out? I don't see what it adds to the MADR

Reply:
Maybe make an issue for it

Reply:
https://github.com/kumahq/kuma/issues/6244